### PR TITLE
Initialize OpenGL from C++ using EGL on Android

### PR DESCRIPTION
**Pull request by hrydgard**
One small part of #8272, but done correctly :)

Lets us avoid some error-prone Java code for initializing OpenGL. This also makes it possible for us to initialize desktop GL (non-ES) on Android in the future, which is potentially useful on some hardware, although ES has been catching up lately.

May look at NativeActivity again later but for now this should do.

Also cleans up DPI and Android HW-scaler management, and lets us stay fully in C++ in the render loop on Android, no dalvik involvement at all.

### CODE REVIEW

**Comment by hrydgard** [comment]
This appears to run fine on all devices I have, but rotation is mishandled on some older devices. Fixing that.

**Comment by hrydgard** [comment]
OK, this now works fine on all my devices. @unknownbrackets wanna have a look? otherwise I'll merge later, if it passes travis.

File: Common/GL/GLInterface/EGL.cpp [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+		// attributes for a visual in RGBA format with at least
+		// 8 bits per color
+		int attribs[] = {
+			EGL_RED_SIZE, 8,
+			EGL_GREEN_SIZE, 8,
+			EGL_BLUE_SIZE, 8,
+			EGL_ALPHA_SIZE, 8,
+			EGL_DEPTH_SIZE, 16,
+			EGL_STENCIL_SIZE, 8,
+			EGL_RENDERABLE_TYPE, renderable_type,
```
Comment (unknownbrackets) [inline Common/GL/GLInterface/EGL.cpp]: Per that nice article you linked to, do we want `EGL_TRANSPARENT_TYPE` and `EGL_SURFACE_TYPE` here?

-[Unknown]

File: android/jni/app-android.cpp [hunk 2]
```diff
@@ -547,3 +465,210 @@ extern "C" void Java_org_ppsspp_ppsspp_NativeApp_sendMessage(JNIEnv *env, jclass
...
+	bool use565 = false;
+	switch (backbuffer_format) {
+	case 4:  // PixelFormat.RGB_565
+		use565 = true;
+		break;
+	}
+
+	if (!gl->Create(wnd, false, use565)) {
+		ELOG("EGL creation failed");
+		// TODO: What do we do now?
```
Comment (unknownbrackets) [inline android/jni/app-android.cpp]: I guess we explode.  If this just returns, it seems like it'll never die I guess?

Should we `exit(1)` at least?

-[Unknown]

File: android/src/org/ppsspp/ppsspp/NativeActivity.java [hunk 3]
```diff
@@ -309,71 +286,66 @@ public void checkForVibrator() {
...
 	}
 
-	// Override this to scale the backbuffer (use the Android hardware scaler)
-	public void getDesiredBackbufferSize(Point sz) {
-		sz.x = 0;
-		sz.y = 0;
+	private Runnable mEmulationRunner = new Runnable() {
+		@Override
+		public void run() {
+			// Bit of a hack - loop until onSurfaceCreated succeeds.
```
Comment (unknownbrackets) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: Should we just defer this (set a flag) to when the surface has been created?

-[Unknown]

File: android/src/org/ppsspp/ppsspp/NativeActivity.java [hunk 4]
```diff
@@ -309,71 +286,66 @@ public void checkForVibrator() {
...
 
+    	updateDisplayMetrics(null);
+
 		if (!initialized) {
 			Initialize();
 			initialized = true;
 		}
+
+		// OK, config should be initialized, we can query for screen rotation.
+		if (Build.VERSION.SDK_INT >= 9) {
```
Comment (unknownbrackets) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: Any reason why not `Build.VERSION_CODES.GINGERBREAD`?

-[Unknown]

File: android/src/org/ppsspp/ppsspp/NativeActivity.java [hunk 5]
```diff
@@ -381,8 +353,62 @@ public void onCreate(Bundle savedInstanceState) {
...
+
+	@Override
+	public void surfaceChanged(SurfaceHolder holder, int format, int width, int height)
+	{
+		Log.w(TAG, "Surface changed. Resolution: " + width + "x" + height + " Format: " + format);
+		// Make sure we have fresh display metrics so the computations go right.
+		// This is needed on some very old devices, I guess event order is different or something...
+		Point sz = new Point();
+        updateDisplayMetrics(sz);
+        NativeApp.backbufferResize(width, height, format);
```
Comment (unknownbrackets) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: nit: Mixed tabs and spaces.

-[Unknown]

File: android/src/org/ppsspp/ppsspp/NativeActivity.java [hunk 6]
```diff
@@ -309,71 +286,66 @@ public void checkForVibrator() {
...
-        //if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
-        
-        // On some (especially older devices), things blow up later (EGL_BAD_MATCH) if we don't set the format here,
-        // if we specify that we want destination alpha in the config chooser, which we do.
-        // http://grokbase.com/t/gg/android-developers/11bj40jm4w/fall-back
-        
-        
-        // Needed to avoid banding on Ouya?
-        if (Build.MANUFACTURER == "OUYA") {
-        	mGLSurfaceView.getHolder().setFormat(PixelFormat.RGBX_8888);
```
Comment (unknownbrackets) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: Are we thinking the new chooser logic won't need this?

-[Unknown]

**Comment by unknownbrackets** [comment]
Looks pretty good to me.

-[Unknown]

File: android/src/org/ppsspp/ppsspp/NativeActivity.java (hunk 6 above)
Comment (hrydgard) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: I hope so, I realized I forgot to dig out the Ouya so I'll check and see if I still get banding.

File: android/src/org/ppsspp/ppsspp/NativeActivity.java (hunk 5 above)
Comment (hrydgard) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: Yeah, I'll have see if I can configure Eclipse better.. for some reason this always turns into a mess.

File: android/src/org/ppsspp/ppsspp/NativeActivity.java (hunk 4 above)
Comment (hrydgard) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: Should just remove the check instead, we don't support pre-gingerbread anyway.

File: Common/GL/GLInterface/EGL.cpp (hunk 1 above)
Comment (hrydgard) [inline Common/GL/GLInterface/EGL.cpp]: Right, forgot to specify it here in Detect too. Will do.

File: android/jni/app-android.cpp (hunk 2 above)
Comment (hrydgard) [inline android/jni/app-android.cpp]: Reporting can work, I think. But yeah, a Java alert dialog is probably the way to go, will require a little bit of plumbing to pass back an error code.

File: android/src/org/ppsspp/ppsspp/NativeActivity.java (hunk 3 above)
Comment (hrydgard) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: Might be better, yes ..

File: android/src/org/ppsspp/ppsspp/NativeActivity.java (hunk 6 above)
Comment (hrydgard) [inline android/src/org/ppsspp/ppsspp/NativeActivity.java]: With or without this, Ouya has a strangely banded output on my TV, although not as banded as 16-bit would be. So I'm skipping this.

**Pull request merged**